java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -386,16 +421,63 @@ public void runTask(final ShardedJobRunId jobId, final IncrementalTaskId taskId,
-        IncrementalTaskState taskState =
-          getAndValidateTaskState(lockAcquisition, taskId, sequenceNumber, jobState);
-        if (taskState == null) {
-          // some sort of error code happened
-
-          // seems like getAndValidationTaskState has potential side-effects, which need to be committed
-          lockAcquisition.commit();
-          return null;
+        IncrementalTaskState taskState = lookupTaskState(lockAcquisition, taskId);
+        IncrementalTaskStateValidationError validationError = validateTaskState(taskState, sequenceNumber, jobState);
+        if (validationError != null) {
+          switch (validationError) {
+            case TASK_GONE:
+              log.info(taskId + ": Task is gone, ignoring runTask call.");
+              lockAcquisition.commit();
+              //throw, so we re-try a few moments later and see if taskState has been updated
+              throw new IllegalStateException("Task is gone");
+
+            case TASK_NO_LONGER_ACTIVE:
+              log.info(taskId + ": Task is no longer active, ignoring runTask call.");
+              lockAcquisition.commit();
+              return null; //we're done here
+
+            case JOB_NO_LONGER_ACTIVE:
+              taskState.setStatus(new Status(StatusCode.ABORTED));
+              log.info(taskId + ": Job no longer active: " + jobState + ", aborting task.");
+
+              //TODO: has side-effects (enqueuing things)
+              updateTask(lockAcquisition, jobState, taskState, null, false);
+              lockAcquisition.commit();
+              return null; // we're done here
+
+            case TASK_STATE_FROM_PAST:
+              log.warning(taskId + " sequenceNumber=" + sequenceNumber + " : Task state is from the past: " + taskState);
+              // presumably we are reading an old state
+              // task to execute sequenceNumber was enqueued, but state in datastore does not yet reflect it
+              // this can happen now because we no longer have transactions across Cloud Tasks + Datastore
+              // we should not proceed with this state, throw an IllegalStateException to force retry
+              lockAcquisition.commit();
+              //throw, so we re-try a few moments later and see if taskState has been updated
+              throw new IllegalStateException("Task state is from the past");
+
+            case TASK_ALREADY_COMPLETED:
+              log.info(taskId + ": Task sequence number " + sequenceNumber + " already completed: "
+                + taskState);
+              lockAcquisition.commit();
+              return null; // we're done here
+
+            case LOCK_HELD_BY_OTHER_EXECUTION:
+              //TODO: has side-effects (enqueuing things); essentially underneath makes decisions that better to wait by enqueuing fresh Cloud Task
+              // than waiting in this thread ... refactor??
+              handleLockHeld(lockAcquisition, taskId, jobState, taskState);
```

Comment:
this ends up in a potentially LONG chain of calls, as it checks slice/shard retry policies in the abandon lock case, and potentially triggers those.